File: tests/fixtures/vulnerable.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/md5.h>

// Buffer overflow: gets()
void read_input(char *buf) {
    gets(buf);
}

// Buffer overflow: strcpy without bounds
void copy_name(char *dest, const char *src) {
    strcpy(dest, src);
}

// Buffer overflow: sprintf without bounds
void format_msg(char *buf, const char *name) {
    sprintf(buf, "Hello, %s!", name);
}

// Buffer overflow: scanf with unbounded %s
void scan_name(char *buf) {
    scanf("%s", buf);
}

// Format string vulnerability
void log_message(const char *msg) {
    printf(msg);
}

// Command injection via system() with user input
void run_cmd(const char *input) {
    char cmd[256];
    sprintf(cmd, "echo %s", input);
    system(cmd);
}

// Hardcoded secret
const char *api_key = "sk-secret-1234567890abcdef";

// Insecure random
int generate_token() {
    srand(time(NULL));
    return rand();
}

// Weak crypto: MD5
void hash_data(const unsigned char *data, size_t len) {
    unsigned char digest[16];
    MD5(data, len, digest);
}

File: tests/fixtures/vulnerable.rb
require 'digest'

class VulnerableController < ApplicationController
  # SQL injection via string interpolation
  def show
    user = User.where("name = '#{params[:name]}'")
  end

  # XSS via html_safe
  def render_content
    content = params[:body].html_safe
  end

  # Command injection via backtick with interpolation
  def run_cmd
    result = `ls #{params[:dir]}`
  end

  # Command injection via system
  def execute
    system("echo #{params[:msg]}")
  end

  # Hardcoded secret
  SECRET_TOKEN = "my_super_secret_token_12345"

  # Mass assignment via permit(:all)
  def create
    User.create(params.require(:user).permit(:all))
  end

  # Path traversal via File.read with params
  def download
    File.read(params[:path])
  end

  # Insecure random
  def generate_token
    rand(1000000)
  end

  # Weak crypto: MD5
  def hash_password(pwd)
    Digest::MD5.hexdigest(pwd)
  end

  # Open redirect
  def redirect
    redirect_to params[:url]
  end

  # Eval injection
  def dynamic_eval
    eval(params[:code])
  end
end
